### ISSUE TITLE
Improve submission wizard accessibility and usability

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7191]
## Description

**Describe the problem you would like to solve**
The submission wizard relies on older forms components that do not meet accessibility requirements. Also, the submission wizard is not as easy to use as it could be, because the steps are not clearly explained and the wizard-like UI looks more like tabs than a step-by-step process.

**Describe the solution you'd like**
Rebuild the submission wizard using the new forms components. Build it to meet accessibility requirements for use by keyboard or mouse, with or without assistive technology. Redesign the submission wizard to give a clear indication of the step-by-step process.

**Who is asking for this feature?**
This rebuild is required before we can commission an accessibility audit of the submission wizard.

**Additional information**
We hope this work will incorporate additional features requested in the [submission wizard project](https://github.com/pkp/pkp-lib/projects/20), but we can't yet commit to this within 3.4's timeline. 

## Mockups

@israelcefrin, @ewhanson and I have worked together to identify ways to improve our existing submission wizard. Israel conducted a benchmark analysis of other submission systems. And we completed a card sorting exercise together to figure out how to break the wizard down into different steps. 

The mockups below were made based on this work, as well as some of the technical requirements of the wizard. Israel will conduct UX walkthrough testing with these mockups and we are now looking for feedback from our community.

The mockups below **do not include all possible submission data**. They show the wizard with a typical configuration in OJS, which doesn't require every field. All fields will be accounted for when the new wizard is built.

### Pre-flight

Before the author begins their submission, they are shown the pre-flight screen. This step is necessary to collect some information before the submission begins. The "Before you begin" text will be customizable for each journal.

![Preflight](https://user-images.githubusercontent.com/2306629/126132446-cda57ea1-fba6-418a-ba4e-44fb9e9c83bb.png)

### Step 1: Upload files

Once submission has begun, the user is shown the submission wizard with a step-by-step design. Each step of the wizard has a description to tell the author what they must do on that step. This text can be customized by the journal.

The file upload panel is identical to the new file upload process introduced in 3.3.

![Files](https://user-images.githubusercontent.com/2306629/126138372-407286da-df25-43fa-935a-52f9b39a9d0b.png)

### Step 2: Contributors

Authors are asked to provide details about contributors to the submission. The mockup below shows features that we would like to implement in the future, where author lookup occurs by email or ORCID. However, for 3.4, we are likely to use something more like the existing authors list where each author is added manually.

![Contributors](https://user-images.githubusercontent.com/2306629/126134939-585a84b9-758e-4073-9dd6-2ecf0c9027a5.png)

### Step 3: Details

Authors are asked to provide the key details about their submission. If enabled during submission, references would be collected during this step as well.

OJS / OPS:
![OJS](https://user-images.githubusercontent.com/2306629/126135426-a60a171d-14a4-452c-be3b-2b7c7f76dd8a.png)

OMP:
![OMP](https://user-images.githubusercontent.com/2306629/126135470-9e326ca2-43f2-43d7-90dc-a3276d326882.png)

### Step 4: For the Editors

All other metadata and submission information is provided in this step. We chose to group this information under the heading, "For the Editors", because authors often lack the confidence or ability to provide accurate metadata such as keywords, subjects, etc. 

The help text for this section is written to indicate to the author that this information may not be final, but it should be provided to help the editors manage the submission. We hope that this will reduce an author's fear of making a mistake, and also help them understand why they are being asked for the information.

In OPS, the default text would be changed to reflect authors' greater control over metadata for preprints.

![For the Editor](https://user-images.githubusercontent.com/2306629/126136273-f7ca1863-14f0-4b5c-9f96-029f62929d84.png)

### Step 5: Final Review

The final review step provides an opportunity for the author to review all of their submitted information. Any errors or missing information will be highlighted, and the author can go back to the step with the error to complete the information.

The final review step is also where the author will be asked to confirm any legal terms for the submission.

![Review](https://user-images.githubusercontent.com/2306629/126136761-a6e161da-c9ee-4b71-89f3-fd19ee0a81fc.png)

With errors:
![Review - Errors](https://user-images.githubusercontent.com/2306629/126137778-6ffffae7-b0df-4c6a-9051-cd7b5b0b4d67.png)

### Autosave

When viewing the submission wizard, all changes will be autosaved. The author does not need to click the save button to have their changes saved, and the UI will show them when the last change was saved.

![Last saved](https://user-images.githubusercontent.com/2306629/126136996-72e95431-bf63-4d6a-87d8-1c4c0afd9d75.png)

### Save for Later

Authors can save their submission and come back to it later at any time. When they click the Save for Later button, they'll be taken to a screen that confirms the submission was saved. They will be shown a link to access their submission again, and this link will be emailed to them.

![Saved for Later](https://user-images.githubusercontent.com/2306629/126137188-f7257780-373c-4405-b350-ff8956dc1bba.png)

### Phone/Mobile Layout

On a small screen, the wizard will be collapsed into an accordion that shows the current step. It can be expanded to show all steps and move between them.

![Layout](https://user-images.githubusercontent.com/2306629/126137452-498028bb-31dc-4505-9941-ab19cd44effc.png)

Wizard expanded:
![Wizard Steps Expanded](https://user-images.githubusercontent.com/2306629/126137456-c28b7ff8-d0e1-4dee-9296-c8b422c353a0.png)

## Pull Requests
See [comment](https://github.com/pkp/pkp-lib/issues/7191#issuecomment-1283823331).
- https://github.com/pkp/ui-library/pull/221
- https://github.com/pkp/pkp-lib/pull/8349
- https://github.com/pkp/ojs/pull/3579
- https://github.com/pkp/omp/pull/1220
- https://github.com/pkp/ops/pull/371

**Comment by asmecher** [comment]
For the ORCID invite, I envision a "ghost" entry that lists unconfirmed invitees along with conventionally entered authors. Otherwise users will probably try an ORCID invite, see no immediate effect, then enter the record manually, leading to a duplicate.

**Comment by NateWr** [comment]
:+1: This part will need a lot of fleshing out. I don't plan to do it for 3.4, but I envision it working alongside a confirmation workflow for authors that includes ORCID authentication and CRediT assignment. Broadly speaking:

1. Submitting only author enters email or ORCID for each author during submission.
2. Each author is emailed to notify them of the submission. They are asked to click a link to confirm they are authors.
3. When they click the link, they are asked to provide full details (name, affiliation, CRediT roles) and to authorize their ORCID.

This would reduce the burden on the submitting author, ensure we're getting author information "from the horse's mouth", and provide built-in protection for predatory authorship behaviour (authors listed that shouldn't be, disputes over CRediT role assignment, etc).

**Comment by asmecher** [comment]
Sounds great, though I do think we'd need to check with the user community in some form before dropping support for submitting authors just filling in what they want, since that's the long-standing habit. I can ping the TC as a community proxy when we're ready to start thinking about that. And I think we should integrate the ORCID profile plugin into the core codebase before doing the ORCID aspects of what you've proposed.

**Comment by alexxxmendonca** [comment]
We should be mindful that as best practice, ORCID no longer recommends free-text input of ORCID IDs.

Adding the email should be enough and ideally the system would automatically send an e-mail to that address requesting the ORCID association.

**Comment by aguen** [comment]
Hi Nate, thank you for working on this! I like the "Save for Later" option (including email with a resume link), also the "Final Review" page (including legal terms).

One question on the "For the Editors" page: Will this page be configurable, meaning that journal managers are able to choose which metadata fields are presented to submitting authors and which not? In addition, a possibility to add custom fields would be great. (Perhaps the review form editor could be re-used for this?)

**Comment by NateWr** [comment]
> We should be mindful that as best practice, ORCID no longer recommends free-text input of ORCID IDs.

This would kick off an authentication process, so it shouldn't be subject to the problems of unauthenticated or inaccurate ORCID IDs.

> One question on the "For the Editors" page: Will this page be configurable, meaning that journal managers are able to choose which metadata fields are presented to submitting authors and which not? In addition, a possibility to add custom fields would be great. (Perhaps the review form editor could be re-used for this?)

The same metadata configuration options will exist. So things like keywords, subjects, references, etc can be included or not included in the submission wizard. We don't plan to support custom fields in the next major version (3.4), but it should be easier to add support for this feature in later versions. In the meantime, plugins can be written with custom fields.

cc @ViRuSTriNiTy if I recall, you wrote a plugin to add custom fields to the submission wizard. This work will impact your plugin, but it should make it easier to add your custom fields. 

> (BTW: It is good to NOT use the technical term "Metadata" here!)

:raised_hands:

**Comment by NateWr** [comment]
@israelcefrin conducted usability tests on these mockups, with the following recommendations.

* The page title "Make a Submission" should include the current step, such as "Make a Submission: Upload Files".
* When an error appears in the review step, add a link to the message at the top of the screen that will jump down to the first error.
* For each error on the review step, add a link to the step to fix the error. For example, in teh mockup above. The abstract error might read: "The abstract must be at least 200 words. [Edit](#)"
* Add the Save for Later button to the top of the page, in addition to the button at the bottom of the page, so that it is easier to find.
* On the pre-flight step, the section selection should be a radio button.
* The description text in the Upload Files step should tell the author that they will be able to write a comment to the editor at a later step, where they can describe the files that they have uploaded.
* All mandatory fields should show that they are required.

In addition, we had some feedback on the contributors step. Unfortunately, we won't be able to implement the contributors mockup for 3.4, so I have filed that as a separate issue for later development: https://github.com/pkp/pkp-lib/issues/7223

**Comment by ronste** [comment]
Hi @NateWr ,

I very much like the new layout of the submission process as you suggested. My question would be what exactly happens when you press the submit button? I presume mails will be send to the author and the journal acknowledging the submission.

Will these mails contain information about the terms and conditions the author agreed to? Or do you have plans to document the legal act of agreeing to the terms and conditions in any other way (e.g. logging)?

The context of my question is of cause our previous discussion: https://forum.pkp.sfu.ca/t/documentation-of-the-state-of-submission-checkboxes-at-time-of-submission/62631/8.

Best wishes,
Ronald

**Comment by NateWr** [comment]
Hi @ronste,

> My question would be what exactly happens when you press the submit button? I presume mails will be send to the author and the journal acknowledging the submission.

The email notifications will be the same as before. There are some complex rules about who gets what, but we don't currently plan to change this for 3.4.

> Will these mails contain information about the terms and conditions the author agreed to? Or do you have plans to document the legal act of agreeing to the terms and conditions in any other way (e.g. logging)?

We've discussed this need at https://github.com/pkp/pkp-lib/issues/7062 (and slightly different, but similar need, related to privacy consent at https://github.com/pkp/pkp-lib/issues/6180). In the mockups above, I've included a checkbox in the final step that would address this, but I'm not yet sure that it will be implemented for 3.4. I hope to get to it, but we may run out of time.

**Comment by NateWr** [comment]
While working on this, we will build out our submissions API for creating and editing submissions. We will have to restrict authors from editing submission properties, like the `stageId` or `status`. However, should we also restrict managers from editing some of these properties through the API?

Here are some of the properties we might want to restrict:

| Property | Description |
| --- | --- |
| `contextId` | Never allow this to be modified through the API. Make a separate API endpoint for migrating submissions between contexts if that is needed. |
| `currentPublicationId` | Only allow this to be modified by creating, deleting, publishing or unpublishing publications? |
| `dateLastActivity`, `lastModified` | Never allow these to be modified through the API. These should be set internally when actions are taken. |
| `dateSubmitted` | Only allow editors to set this when a new submission is created (for example, when importing back content). Once it is set, it can't be changed through the API. |
| `locale` | Changing the locale can cause problems with missing multilingual data... I'm not sure how to handle this. |
| `stageId`, `status` | Only these to be modified through requests to the editorial decisions API? |

Can I get some feedback on this from the community who might use the API? I want the API to be as useful as possible, but if we let editors do _anything_ we'll probably create a lot of circumstances where data gets confusing -- for example, a `STATUS_PUBLISHED` in the submission stage.

**Comment by NateWr** [comment]
Regarding the API restrictions mentioned in [this comment](https://github.com/pkp/pkp-lib/issues/7191#issuecomment-1171014633). I've disabled editing all of the props mentioned through the add and edit API endpoints for submissions except the `locale`. We still need authors and editors to be able to switch this, and won't have a dedicated API endpoint for now.

**Comment by NateWr** [comment]
The following PRs implement the new submission wizard.

PRs:
- https://github.com/pkp/ui-library/pull/221
- https://github.com/pkp/pkp-lib/pull/8349
- https://github.com/pkp/ojs/pull/3579
- https://github.com/pkp/omp/pull/1220
- https://github.com/pkp/ops/pull/371

@asmecher can you do a code review? I recommend just installing from scratch, but if you want to use an existing database, you'll want to run all `I7191_` migrations and probably install all email templates again.

Known problems:
- The tests for OJS are passing locally, but seem to be [failing in Travis](https://app.travis-ci.com/github/pkp/ojs/builds/256846261). I'm just getting the message "No output has been received in the last 10m0s" and no error logs. I think this is what you mentioned in the call the other day. Any ideas how to debug this?
- There are no tests for OMP/OPS. They'll need to be converted similar to OJS.
- Some of the emails sent with a new submission are missing email template variables. https://github.com/pkp/pkp-lib/issues/8348

Things to look for in your review:
- I've removed the automated editorial assignments code and changed it to manual assignments by section or category. See the section or category settings.
- I _hope_ the autosave feature is pretty robust. But keep an eye on that as you test the new wizard. You can test a lost connection as well by killing your server while working on the submission wizard. It should show you're disconnected and periodically try to reconnect. Once reconnected, it should replay your autosaves. If you browse away while disconnected, when you come back it should offer to replay your autosaves.

Finally, I've assigned the following issues to 3.4 because they are improvements I'd _really_ like  to make. But if we don't have time for them in the end we can defer/close them:
- https://github.com/pkp/pkp-lib/issues/8350
- https://github.com/pkp/pkp-lib/issues/8351
- https://github.com/pkp/pkp-lib/issues/6241
- https://github.com/pkp/pkp-lib/issues/6306

**Comment by asmecher** [comment]
I've reviewed as far through as I could get this week over here: https://github.com/pkp/pkp-lib/pull/8349#pullrequestreview-1150142997

Mostly the comments there are syntactical or low-level, as that's the stuff I was best able to comment on; TBH I'm still not well oriented enough on VueJS to review the work there properly. But almost instantly I threw a curve-ball at it by changing submission locale at the last moment, and it handled that brilliantly. It looks fantastic, very clean and streamlined, and is clearly written in consideration of some future work that'll be tricky (GDPR, ORCID, etc).

The autosave, pre-submit check, automatic assignment rewrite, events/mailables (kudos also to @Vitaliy-1) are practically science fiction by our historically technically conservative standards. This looks really, really impressive.

I'll complete my first review next week, but given how much new code there is here, and how much scope creep into workflow concerns, maybe we should schedule a quick run-through to make sure I'm able to give it a thorough review. I'll be happy to look at the Travis test failure and porting the tests around too.

**Comment by NateWr** [comment]
Thanks @asmecher! Yeah, I'm really liking the events and listeners model for isolating code in sensibe players (editor assignment, email sending, etc). I've replied to the comments you've made so far. Will look forward to the rest.

**Comment by asmecher** [comment]
@NateWr, I've reviewed this again, done a bunch of testing, and left comments where I spotted anything needing a look. Once those are addressed, please send up the latest code; I'll be ready to take a kick at porting the tests and debugging why the OJS tests are failing, if that's still the case. Thanks, this looks great, and I have additional appreciation e.g. for the autosave feature!

**Comment by NateWr** [comment]
Thanks @asmecher! I've gone through the last round of review comments, rebased the PRs, and pushed up. It looks like Scrutinizer is failing on the doctrine/dbal dependency update:

https://scrutinizer-ci.com/g/pkp/ojs/inspections/c16e3038-14b7-494d-9fbb-b2cf27256f09

OJS tests are still running but they shouldn't be far off passing if they do fail. I'll leave them with you, unless there's anything else you want me to look at.

For convenience, I've put all of the PR links at the bottom of the first comment on this issue.

**Comment by asmecher** [comment]
Scrutinizer appears to be using the latest Composer 1.x, which is long since deprecated. I couldn't find any reference on their website to any configuration options or reasoning for this; if it persists, I think we'll need to disable it and look for other options (which was somewhere on the table anyway -- IIRC this year's dev agendas).

**Comment by asmecher** [comment]
@NateWr, I've made 3 PRs to your i7191_submission_wizard branches:
- https://github.com/NateWr/pkp-lib/pull/13
- https://github.com/NateWr/ojs/pull/7
- https://github.com/NateWr/ops/pull/3 (replaces https://github.com/NateWr/ops/pull/2)

These...
- Complete the OPS test suite (fixing some errors around author data in the last PR)
- Address [your comment](https://github.com/NateWr/pkp-lib/pull/12#pullrequestreview-1184128793) about complicating the file upload Cypress command by breaking the "submit with API" command in to pieces, moving the main call into each application's `commands.js` file. (This will be helpful when it comes to adding OMP tests with e.g. chapter data)
- Make use of [Cypress aliases](https://docs.cypress.io/guides/core-concepts/variables-and-aliases) rather than mixing "what just happened" data into the `data` object

**Comment by asmecher** [comment]
@NateWr, I've added 3 new PRs:
- https://github.com/NateWr/pkp-lib/pull/14 and https://github.com/NateWr/omp/pull/1 fix the first (aclark) OMP submission using the API-based process
- https://github.com/NateWr/ojs/pull/8 fixes an issue you noted in OJS.

I'm going to keep pushing on these but I'm not sure how much I'll get done before I disappear for a bit. Will add further notes here as needed. If you get here first, a review/merge now won't hurt anything.

**Comment by asmecher** [comment]
@NateWr, a bit of a snag with OMP: The series is optional (as are categories), so submissions without series will not by default be assigned to anyone. Therefore submissions that used to get auto-assigned to someone (based on the 1-user-in-user-group thing) now won't have them assigned anywhere, nor will they be able to configure an automatic assignment on anything.

**Comment by asmecher** [comment]
@NateWr, I've edited https://github.com/pkp/pkp-lib/issues/7191#issuecomment-1324353373 to indicate that the PR now includes all the OMP content creation tests -- but they won't pass, at least until https://github.com/pkp/pkp-lib/issues/7191#issuecomment-1324486943 is resolved. I'm OK to have the PR merged into your branch as is, though.

**Comment by asmecher** [comment]
@NateWr, I haven't had enough time this week to do much on the tests, but here are a few more PRs:
- https://github.com/NateWr/ojs/pull/10
- https://github.com/NateWr/omp/pull/2
- https://github.com/NateWr/ops/pull/4

All 3 apps get their own cypress.env.json files in order to work around the weird (and probably outdated) issue with Cypress not consistently pulling the `CYPRESS_baseUrl` environment variables when doing API-based requests.

The OJS PR re-adds a work-around for Touhidur's changes to the language names that I submitted before -- not sure if you removed this part on purpose or it got dropped during rebase? This will make the OJS tests consistent with what's in OMP and OPS (and what the language lists currently look like). Once all tests are passing, I'll ask him to review and fix the issue and remove the work-arounds.

Unfortunately there are some merge conflicts in your code with `main` that have arisen during the week with some of Vitaliy's additions. Again, apologies for not getting this merged this week like I'd hoped -- can you merge these PRs, then rebase against `main`, and I'll pick up the test failures in the second half of the coming week? Thanks!

**Comment by asmecher** [comment]
> I don't think a cypress.env.json file should be committed to the repository.

It's already part of `.gitignore`, so it would need a `git add -f` to be pushed and won't show up e.g. in `git diff` or `git add -u`. I suppose we could generate it on-the-fly, but having it as a basis for a local template and not needing to steward its generation code seem like worthwhile wins to me.

**Comment by jonasraoni** [comment]
I think the problem is that @NateWr probably has a custom `cypress.env.json` (I have one), which he's afraid to be committed by accident, then I think it's better to generate it on-the-fly as well (or keep it under a different name/folder "cypress.prod.env.json", then rename it) 🤔

**Comment by NateWr** [comment]
@asmecher have you done any tests to see how that would work with regular `git pull` and `git rebase` operations? If a local `cypress.env.json` file doesn't conflict at all with the one in the repo, then I have no objections. But I imagine it might cause conflicts or lead to cases where git won't reset the tree because it doesn't want to overwrite local changes.

What about storing the file in `/tools` and copying it into place during the travis tests?

**Comment by NateWr** [comment]
Or yeah, what @jonasraoni suggested about a file name like `cypress.travis.env.json`.

**Comment by asmecher** [comment]
@NateWr and @jonasraoni, I like the idea of using `cypress.travis.env.json` -- it's clearly different than a local `cypress.env.json` but also can serve as a nice template for creating one. I've updated the PRs [listed above](https://github.com/pkp/pkp-lib/issues/7191#issuecomment-1336450285) to achieve that. Please have a look and rebase from `main` when you're able, @NateWr, and I'll take it from there through fixing any remaining tests and merging.

**Comment by NateWr** [comment]
Thanks @asmecher! Ok, I've merged your PRs and rebased everything. Good luck!

**Comment by asmecher** [comment]
@NateWr, I've merged PRs from my forks for pkp-lib, ui-library, ojs, omp, and ops. Those should cover everything in your branches. Is this otherwise ready to be closed?

**Comment by NateWr** [comment]
@emmauhl this will effect documentation. The new submission wizard will look and work differently, and includes autosave and save for later features. The screenshots at the top of the thread are mockups and do not show precisely what the final wizard looks like, so new screenshots will need to be taken. I'll probably be taking lots of these for 3.4 launch communications.

**Comment by asmecher** [comment]
@NateWr, https://github.com/pkp/ojs/blob/247e6081d667648b9014912238cf1ee472b286b8/docs/dev/swagger-source.json#L652 will need updating for the change to the submission_progress column from numeric to string.

**Comment by NateWr** [comment]
This is in my own todo list for docs. I'd rather keep it there, as the API docs have a lot of changes that need to be made and captured in a list of changes from previous versions. If you'd prefer, I can file an issue about the changes.
